x-pack/filebeat/input/websocket/config.go

```diff
@@ -0,0 +1,152 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package websocket
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"regexp"
+	"time"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type config struct {
+	// Program is the CEL program to be run for each polling.
+	Program string `config:"program"`
+	// Regexps is the set of regular expression to be made
+	// available to the program.
+	Regexps map[string]string `config:"regexp"`
+	// State is the initial state to be provided to the
+	// program. If it has a cursor field, that field will
+	// be overwritten by any stored cursor, but will be
+	// available if no stored cursor exists.
+	State map[string]interface{} `config:"state"`
+	// Auth is the authentication config for connection
+	Auth authConfig `config:"auth"`
+	// Resource
+	Resource *ResourceConfig `config:"resource" validate:"required"`
+	// Redact is the debug log state redaction configuration.
+	Redact *redact `config:"redact"`
+}
+
+type ResourceConfig struct {
+	URL   *urlConfig  `config:"url" validate:"required"`
+	Retry retryConfig `config:"retry"`
+}
+
+type redact struct {
+	// Fields indicates which fields to apply redaction to prior
+	// to logging.
+	Fields []string `config:"fields"`
+	// Delete indicates that fields should be completely deleted
+	// before logging rather than redaction with a "*".
+	Delete bool `config:"delete"`
+}
+
+type authConfig struct {
+	// Api-Key to use for authentication.
+	ApiKey *apiKeyConfig `config:"api_key"`
+	// Baerer token to use for authentication.
+	BearerToken string `config:"bearer_token"`
+	// Basic auth token to use for authentication.
+	BasicToken string `config:"basic_token"`
+}
+
+type apiKeyConfig struct {
+	// Api-Key to use for authentication.
+	Header string `config:"header"`
+	Value  string `config:"value"`
+}
+type urlConfig struct {
+	*url.URL
+}
+
+func (u *urlConfig) Unpack(in string) error {
+	parsed, err := url.Parse(in)
+	if err != nil {
+		return err
+	}
+	u.URL = parsed
+	return nil
+}
+
+func (c config) Validate() error {
+	if c.Redact == nil {
+		logp.L().Named("input.websocket").Warn("missing recommended 'redact' configuration: " +
+			"see documentation for details: https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-websocket.html#_redact")
+	}
+	_, err := regexpsFromConfig(c)
+	if err != nil {
+		return fmt.Errorf("failed to check regular expressions: %w", err)
+	}
+
+	var patterns map[string]*regexp.Regexp
+	if len(c.Regexps) != 0 {
+		patterns = map[string]*regexp.Regexp{".": nil}
+	}
+	_, _, err = newProgram(context.Background(), c.Program, root, patterns, logp.L().Named("input.websocket"))
+	if err != nil {
+		return fmt.Errorf("failed to check program: %w", err)
+	}
+	return nil
+}
+
+type retryConfig struct {
```

Comment:
I don't think we need this since we are not an HTTP client.

Reply:
I was thinking of keeping an auto reconnect mechanism if the connection breaks for some reason.

Reply:
Let's add it in when the time arises, the model may not be the same.